src/backend/distributed/commands/cascade_citus_table_function.c

```diff
@@ -0,0 +1,294 @@
+static void DropRelationIdListForeignKeys(List *relationIdList);
+static void DropRelationForeignKeys(Oid relationId);
+static List * GetRelationDropFkeyCommands(Oid relationId);
+static char * GetDropFkeyCascadeCommand(Oid relationId, Oid foreignKeyId);
+static void ExecuteCitusTableFunctionForRelationIdList(List *relationIdList,
+													   ExecuteCitusTableFunctionFlag
+													   executeCitusTableFunctionFlag);
+static CitusTableFunc GetCitusTableFunction(ExecuteCitusTableFunctionFlag
+											executeCitusTableFunctionFlag);
+
+
+/*
+ * ExecuteCitusTableFunctionCascade executes citus table function specified
+ * by ExecuteCitusTableFunctionFlag argument on each relation that relation
+ * with relationId is connected via it's foreign key graph, which includes
+ * input relation itself.
+ * Also see ExecuteCitusTableFunctionFlag enum definition for supported
+ * citus table functions.
+ */
+void
+ExecuteCitusTableFunctionCascade(Oid relationId, LOCKMODE lockMode,
+								 ExecuteCitusTableFunctionFlag
+								 executeCitusTableFunctionFlag)
+{
+	List *fKeyConnectedRelationIdList = GetForeignKeyConnectedRelationIdList(relationId);
+	LockRelationsWithLockMode(fKeyConnectedRelationIdList, lockMode);
+
+	/*
+	 * Our foreign key subgraph can have distributed tables which might already
+	 * be modified in current transaction. So switch to sequential execution
+	 * before executing any ddl's to prevent erroring out later in this function.
+	 */
+	EnsureSequentialModeForCitusTableCascadeFunction();
+
+	/* store foreign key creation commands before dropping them */
+	List *fKeyCreationCommands =
+		GetFKeyCreationCommandsForRelationIdList(fKeyConnectedRelationIdList);
+
+	DropRelationIdListForeignKeys(fKeyConnectedRelationIdList);
+	ExecuteCitusTableFunctionForRelationIdList(fKeyConnectedRelationIdList,
+											   executeCitusTableFunctionFlag);
+
+	/* now recreate foreign keys on tables */
+	ExecuteAndLogDDLCommandList(fKeyCreationCommands);
+}
+
+
+/*
+ * LockRelationsWithLockMode sorts given relationIdList and then acquires
+ * specified lockMode on those relations.
+ */
+static void
+LockRelationsWithLockMode(List *relationIdList, LOCKMODE lockMode)
+{
+	Oid relationId;
+	relationIdList = SortList(relationIdList, CompareOids);
+	foreach_oid(relationId, relationIdList)
+	{
+		LockRelationOid(relationId, lockMode);
+	}
+}
+
+
+/*
+ * EnsureSequentialModeForCitusTableCascadeFunction switches to sequential
+ * execution mode if possible. Otherwise, errors out.
+ */
+static void
+EnsureSequentialModeForCitusTableCascadeFunction(void)
+{
+	if (!IsTransactionBlock())
+	{
+		return;
+	}
+
+	if (ParallelQueryExecutedInTransaction())
+	{
+		ereport(ERROR, (errmsg("cannot execute table udf because there was a parallel "
+							   "operation on a distributed table in transaction"),
+						errhint("Try re-running the transaction with "
+								"\"SET LOCAL citus.multi_shard_modify_mode TO "
+								"\'sequential\';\"")));
```

Comment:
some more details here ?

also, I think we should unify those EnsureSequentialModeForXXX functions, I will open a seperate issue for that